rgw: bucketmv

mdw-at-linuxbox opened the pull request:
When converting from a global implicit tenant setup to a multi-tenancy setup, old buckets become inaccessible.  Jewel had a somewhat strange compromise setup for this, which had several strange properties.  Modern versions of ceph have more uniform properties, but require that buckets be somehow moved to the current data layout.  This series of patches provides an administrative way to move buckets between tenant.  As a  bonus this can also be used to rename buckets.

This is a slightly complicated change, so instead of one giant scary commit, I have,

8297f4cfc7 rgw: making implicit_tenants backwards compatible.
e9cf92a3d6 Add several types to ceph-dencoder.
d6f71e0a50 Add "--bucket-new-name" option to radosgw-admin.
bb3ba20e85 rgw: bucket link: Add ability to name bucket w/ different tenant.
fa40823d4a rgw: bucket link: simplify use of get bucket info.
a847ace94a rgw: bucket link: use data from bucket_info to rewrite bucket_endpoint.
4499845d35 rearrange / simplify RGWBucket::link logic - start bucket move support
d8efa3b2c5 rgw: bucket link: base "bucket move" (tenant id only)
b33d851d72 rgw: bucket link: "bucket move"; handle bucket names too.
0208764a57 rgw: bucket link: "bucket move" documentation changes

Note that 8297f4cfc7 is (or should be the same) as PR#22378
e9cf92a3d6 (dencoder) isn't strictly required, but made debugging this a lot easier with the ability to more easily print out rados objects associated with rgw buckets.  I see no reason not to include this.

The remaining patches incrementally build "bucket link" functionality; note that some patches overwrite changes from earlier patches - I'm not sure this will make it as easy to review this as I had hoped.

Several things are not here because they appear to not be necessary: this does not loop and rewrite object acls in a bucket.  Swift at least doesn't appear to care.  Doesn't rewrite tail placement; doesn't seem to break anything.  No real provisions to lock the bucket.  Resharding a bucket seems like the most likely competitor; its locking logic doesn't appear to be compatible to this.  If you think I'm missing functionality please if possible provide testable examples - if I can't test it I can't know that my code is right.

So possible things left: squash patches?  create more tracker issues and reference them?  Other things people identify here?

vumrao commented:
I have created this tracker - http://tracker.ceph.com/issues/35885 for this issue.

rzarzynski commented on src/rgw/rgw_auth.cc:
```diff
@@ -441,11 +492,16 @@ void rgw::auth::RemoteApplier::load_acct_info(RGWUserInfo& user_info) const
+      return;
   }
 
+  ldout(cct, 0) << "NOTICE: couldn't map swift user " << acct_user << dendl;
```
Surely *swift user* isn't a new thing but might be worth fixing.

rzarzynski commented on src/rgw/rgw_auth.cc:
```diff
@@ -431,8 +474,16 @@ void rgw::auth::RemoteApplier::load_acct_info(RGWUserInfo& user_info) const
+   * only use the identifier space that would be used if the id were
+   * to be created. */
+
+  if (split_mode && !implicit_tenant)
```
```cpp
   *
   * The truth table for the loads:
   *   split_mode = false, implicit_tenant = false -> (tenanted_uid IF acct_user.tenant.empty()) || acct_user
   *   split_mode = false, implicit_tenant = true  -> (tenanted_uid IF acct_user.tenant.empty()) || acct_user
   *   split_mode = true,  implicit_tenant = false -> acct_user
   *   split_mode = true,  implicit_tenant = true  -> tenanted_uid IF acct_user.tenant.empty() */
```

rzarzynski commented on src/rgw/rgw_auth.h:
```diff
@@ -396,22 +433,26 @@ class RemoteApplier : public IdentityApplier {
 
   const AuthInfo info;
-  const bool implicit_tenants;
+  rgw::auth::ImplicitTenants& implicit_tenant_context;
```
Hmm, does `RemoteApplier` really need this bit as non-`const`?

rzarzynski commented on src/tools/ceph-dencoder/types.h:
```diff
@@ -440,6 +446,7 @@ TYPE(RGWUserInfo)
+TYPE(RGWBucketEntryPoint)
 TYPE(RGWUploadPartInfo)
 TYPE(rgw_obj)
 
```
ACK.

rzarzynski commented on src/test/cli/radosgw-admin/help.t:
```diff
@@ -175,6 +175,8 @@
+                               for bucket link: optional new name
      --shard-id=<shard-id>     optional for: 
                                  mdlog list
                                  data sync status
```
ACK.

rzarzynski commented on src/rgw/rgw_bucket.cc:
```diff
@@ -764,25 +764,38 @@ static void set_err_msg(std::string *sink, std::string msg)
 
+  // split possible tenant/name
+  auto pos = bucket_name.find('/');
+  if (pos != boost::string_ref::npos) {
```
Please consider switching to `std::string::npos` as:

```cpp
/*
 * A simple wrapper class for administrative bucket operations
 */

class RGWBucket
{
  // ...
  std::string bucket_name;
```

rzarzynski commented on src/rgw/rgw_auth.cc:
```diff
@@ -385,7 +385,47 @@ void rgw::auth::RemoteApplier::to_str(std::ostream& out) const
+{
+  static const char *keys[] = {
+    "rgw_keystone_implicit_tenants",
+  NULL };
```
nit: `NULL` -> `nullptr`.

mattbenjamin commented on src/rgw/rgw_bucket.cc:
```diff
@@ -810,13 +825,19 @@ int RGWBucket::link(RGWBucketAdminOpState& op_state, std::string *err_msg)
   }
 
   string bucket_id = op_state.get_bucket_id();
+#if 0
```
what is up with the #if 0 block?

mattbenjamin commented on src/rgw/rgw_bucket.cc:
```diff
@@ -765,7 +765,7 @@ static void set_err_msg(std::string *sink, std::string msg)
 
 int RGWBucket::init(RGWRados *storage, RGWBucketAdminOpState& op_state,
-	std::string *err_msg)
+	std::string *err_msg, map<string, bufferlist> *pattrs)
```
why pointer to pattrs (and err_msg, for that matter)--could these be reference/& arguments?

mattbenjamin commented on src/rgw/rgw_bucket.cc:
```diff
@@ -802,96 +831,139 @@ int RGWBucket::init(RGWRados *storage, RGWBucketAdminOpState& op_state)
-  RGWBucketInfo bucket_info;
+  map<string, bufferlist>::iterator aiter = attrs.find(RGW_ATTR_ACL);
+  if (aiter == attrs.end()) {
+    // XXX why isn't this an error?  mdw 20180825
```
thanks, robin

mattbenjamin commented:
http://pulpito.ceph.com/mbenjamin-2018-09-11_07:12:34-rgw-wip-rgw-bucketmv-distro-basic-smithi/

cbodley commented on src/rgw/rgw_bucket.cc:
```diff
@@ -802,96 +831,139 @@ int RGWBucket::init(RGWRados *storage, RGWBucketAdminOpState& op_state)
+  // XXX I am not convinced this is at all necessary; but previous
+  //	versions would have found and copied VERSION_ATTR so I will
+  //	do likewise for now...  mdw 20180825
+  auto version_iter = attrs.find(VERSION_ATTR);
```
i agree that this part isn't necessary. since we're creating a new object, we should throw out anything related to cls_version tracking of the old object

robbat2 commented:
stalebot: please keep this open

mattbenjamin commented:
@robbat2 thanks for unstaling;  @mdw-at-linuxbox please prioritize a rebase for this; I'll work with @mdw-at-linuxbox to get this a standalone suite run--the prior version had passed

mdw-at-linuxbox commented:
github apparently won't let me respond individually above.  But:
1, added tracker issues to commits.
2, new extra commit for changes to adapt to newer master.
3, replaced hated !! with static_cast<bool>.
4, don't want to do atomics for configuration - I don't think in this case the difference will even be perceptible in human terms - and atomics incur significance overhead the other 99.9999% time.
5, const reference - no point.  references are constant already.
6, -1: replaced some boost::stirng_ref::npos with std::string:npos
7, NULL->nullptr
8, can't use reference for *pattrs.  Well, could, but this would obfuscate the pgm logic, which sets pattrs to one of 2 values in different stmts.
9, added error message to catch error from prehistoric data.
10, removed redundant VERSION_ATTR copy.

mdw-at-linuxbox commented:
github doesn't seem to have absorbed the push I did, and is otherwise acting oddly - I'll try to fix this once things seem more stable.

mdw-at-linuxbox commented:
Got it pushed.  github apparently thought my browser was too old.  :-(

adamemerson commented:
Marcus, please rebase this.

mdw-at-linuxbox commented:
I ran into a problem recreating bz#1669901 - "bucket chown" - which made me very suspicious of this set of commits.  Symptoms included: "old" user could still access objects, "new" owner got permission denied.  Turns out that's all consistent with expected behavior: I was not using "implicit tenants", so the bucket was still accessible in the "old" owner's namespace.  The "new" owner didn't have access, which was the actual feature/bug of 1669901.  So - while there's definitely a problem there, I no longer fear there was damage in the commits here or that they're in any way less functional than what we have for luminous.

I'll rebase and push this later today.

adamemerson commented:
@mdw-at-linuxbox If you would, please ping me via email when you get the rebase for this pushed and I'll shove it into Teuthology and merge it.

mattbenjamin commented:
@mdw-at-linuxbox I don't see a rebase here--did you run into issues after we spoke yesterday?

Matt

mdw-at-linuxbox commented:
I've pushed a rebased version of this.  Rebased on a fresh copy of master, passes my testing.

mdw-at-linuxbox commented:
The assert failure you showed me offline (radosgw_admin_rest.py line 481) looks like something I thought was a previously existing bug.  However, assuming that teuthology test is not supposed to fail, that probably means this is a regression.  I'm trying to compare its behavior to an unmodified build of master (with suiltable instrumentation), but I haven't got a good run of that yet.

pritha-srivastava commented:
This comment is related to BZ 1670217. While running the ACL test against master without the changes in this PR, I could see that the test failed irrespective of whether or not the config rgw keystone implicit tenants is set to true. The reason being that buckets were created under tenant namespace in both the cases. Details are in the BZ.
When the changes in this PR were applied on master and then when I ran the test against it, the test failed when rgw keystone implicit tenants was set to true and passed when the config was absent. It failed because with the config being set to true, the buckets were created under tenant namespace, whereas when the config was absent, buckets were created under global (empty) namespace. 
So is this the correct way to configure rgw, in order to get the test passing in BZ 1670217, and also is this the correct behavior when the config is set to true and when the config is absent?

smanjara commented on src/rgw/rgw_bucket.cc:
```diff
@@ -778,25 +778,38 @@ static void set_err_msg(std::string *sink, std::string msg)
 
   rgw_user user_id = op_state.get_user_id();
   tenant = user_id.tenant;
+  bucket_tenant = tenant;
```
'bucket_tenant' assignment here will pass the new user's tenant to _RGW_Rados::get_bucket_info()_ call in line 796, thus failing with _ret code -2_ in case of original user being untenanted. That's because the 'bucket_tenant' in case of untenanted user will not be updated in line 791 due to absence of tenant/name splitting. This line can be removed safely.

smanjara replied:
Vikhyat pointed out the '/' prefix I missed before the bucket name in the command. Will update the docs.

smanjara commented:
@mdw-at-linuxbox Could you please rebase?

mdw-at-linuxbox commented:
I've pushed a new version against a fresh copy of master, and this copy matches the behavior I see with the corresponding luminous patch.  The rebase I did before this had a lot of weirdness, probably a bad copy of master.

smanjara commented:
@mdw-at-linuxbox Can you please rebase again?

cbodley commented:
as far as i know, everything here merged with https://github.com/ceph/ceph/pull/28813
